### PR TITLE
Hackathon

### PR DESCRIPTION
**Pull request by CedricCabessa**
This branch assemble the work from the last hackathon:
  * More speed
  * A backend

Sorry, @caarmen with the refacto of HTTP request to speed up things, I could'nt easily merge your work. I'll probably do it later.
Anyway this branch is a good basis but need some work to have something mergeable.
Some of known issues:
  * slackbot is broken
  * backend schema is too complicated, which make admin page difficult to use
  * some database ugliness have leaked in the HTTP request engine

### CODE REVIEW

File: web/__init__.py [hunk 1]
```diff
@@ -13,3 +13,25 @@
...
+        super().__init__(**kwargs)
+        self.__token = token
+
+    def https_request(self, req):
+        super().https_request(req)
+        req.add_header('Authorization', 'token %s' % self.__token)
+
+        return req
+
+#FIXME initital migration
```
Comment (calvarez-geny) [inline web/__init__.py]: The solution for the FIXME is here: https://github.com/Genymobile/gm_pr/blob/hackathon-carmen-merge/web/__init__.py

File: web/models.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    """ General settings like url and stuuf like that
+    """
+
+    weburl = models.CharField(max_length=256)
+    github_oauth_token = models.CharField(max_length=256)
+    organization = models.CharField(max_length=256)
+    top_level_url = models.CharField(max_length=256, default=settings.TOP_LEVEL_URL)
+    old_period = models.IntegerField(default=4, null=True, blank=True)
+    slack_settings = models.ForeignKey(SlackSettings, null=True, blank=True)
+    #label_github = models.ManyToManyField(LabelGithub, default=None, blank=True)
```
Comment (calvarez-geny) [inline web/models.py]: Why are these assignments for label_github and feedback_github comented out?  If it's about a warning when starting up the server, this was fixed here:

https://github.com/Genymobile/gm_pr/commit/2ca968f4e326ca6b140340b82b7e3b02714e41ff

Comment (CedricCabessa) [inline web/models.py]: The link is now in the other way
```
class LabelGithub(models.Model):
    general_settings = models.ManyToManyField(GeneralSettings, blank=True)
```

File: common/models.py [hunk 3]
```diff
@@ -32,3 +33,5 @@ def __init__(self, url="", title="", updated_at="", user="",
         self.milestone = milestone
         self.labels = labels
         self.is_old = is_old
+
```
Comment (calvarez-geny) [inline common/models.py]: only whitespace changes in this file.  Maybe revert the changes to prevent conflicts when merging.

File: web/models.py (hunk 2 above)
Comment (calvarez-geny) [inline web/models.py]: Ok. Maybe these commented-out lines could be removed now?

**Comment by sgaland** [comment]
`slackbot is broken` ? :x:

File: bot/slackauth.py [hunk 4]
```diff
@@ -15,11 +15,14 @@
...
 from django.http import HttpResponse
+from web.models import GeneralSettings
+
 
 def isFromSlack(function):
     def __wrap(request, *args, **kwargs):
+        general_settings = GeneralSettings.objects.first()
         if request.GET != None and 'token' in request.GET \
-           and request.GET['token'] == settings.SLACK_TOKEN:
+           and general_settings and request.GET['token'] == general_settings.slack_settings.slack_token:
```
Comment (sgaland) [inline bot/slackauth.py]: What about not returning a 403 when `general_settings` doesn't exist? I think 5xx should be more appropriate.

File: bot/tasks.py [hunk 5]
```diff
@@ -17,20 +17,24 @@
...
+settings_projects = []
+
+#@app.task
 def slack(url, org, weburl, repos, slackurl, channel):
     """ Celery task, use github api and send result to slack
     """
     prf = PrFetcher(url, org, repos)
-    project_list = prf.get_prs()
-    nb_proj = len(project_list)
+    projects = prf.get_prs()
```
Comment (sgaland) [inline bot/tasks.py]: this looks wrong.

File: bot/tasks.py [hunk 6]
```diff
@@ -38,20 +42,21 @@ def slack(url, org, weburl, repos, slackurl, channel):
...
-                txt += " %s review:%d %s:%d %s:%d %s:%d\n" % \
+                txt += " %s review:%d %s:%d %s:%d" % \
                        (pr.user, pr.nbreview,
-                        settings.FEEDBACK_OK['keyword'], pr.feedback_ok,
-                        settings.FEEDBACK_WEAK['keyword'], pr.feedback_weak,
-                        settings.FEEDBACK_KO['keyword'], pr.feedback_ko)
-
+                        settings_projects.FEEDBACK_OK['keyword'], pr.feedback_ok,
+                        settings_projects.FEEDBACK_WEAK['keyword'], pr.feedback_weak)
+                if pr.feedback_ko > 0:
```
Comment (sgaland) [inline bot/tasks.py]: This is awesome, well done.

File: common/prfetcher.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+import re
+
+from django.utils import dateparse
+from django.utils import timezone
+
+from common import paginablejson, models
+from gm_pr.celery import app
+from web.models import ProjectRepository, GeneralSettings, FeedbackGithub, LabelGithub
+
+def is_color_light(rgb_hex_color_string):
```
Comment (sgaland) [inline common/prfetcher.py]: awesome

File: README.md [hunk 8]
```diff
@@ -50,6 +50,8 @@ ProxyPassReverse /gm_pr http://localhost:8000
  * gm_pr/settings.py: this is the standard Django configuration file
  * gm_pr/settings_projects.py: contains everything related to your projects
 
+create superuser
+
```
Comment (calvarez-ov) [inline README.md]: I don't understand this change...  Shouldn't it be included in the manual steps ("hacking" section), after the "migrate" command?  And shouldn't it be:
```
python3 manage.py createsuperuser
```
?

Comment (calvarez-geny) [inline README.md]: I guess we need to update the README in general.  Separate PR maybe.

File: bot/slackauth.py (hunk 4 above)
Comment (calvarez-geny) [inline bot/slackauth.py]: Maybe this is related to this issue: https://github.com/Genymobile/gm_pr/issues/6

File: README.md (hunk 8 above)
Comment (CedricCabessa) [inline README.md]: It is a TODO (forgot to mention)
Yes README need an update, as well as several part of the code. The branch is called "hackaton"

Comment (calvarez-geny) [inline README.md]: Since this branch is called hackathon and has multiple TODOs, breaks things, and has lots of changes... Any hope that my other two PRs, which are quite small and isolated, could be accepted? They should merge onto master without too much problem:

https://github.com/Genymobile/gm_pr/pull/49
https://github.com/Genymobile/gm_pr/pull/47

They add features which could cause a perf hit, but the features which could add a perf hit are only used if you activate them (no perf hit in the default config).

Comment (CedricCabessa) [inline README.md]: Sure. Change the PR or open a new one to target master. If it is mergeable, it's ok for me.
But those feature will need some rewrite to be merged in "hackathon" (which took less than 10sec to display a page versus ~20 before). But I admit it is the guy that break stuff who has the responsability to port the features to the new architecture. And this guy is me (sadly :-))

**Comment by CedricCabessa** [comment]
I think we can kill this now
